### ISSUE TITLE
Feature Request: --redo-ocr or -f with text removal

### ISSUE THREAD

**Issue by julianbarg** [jbarlow83/OCRmyPDF#794]
I am really frustrated by ```-f``` routinely increasing file sizes by a factor of 20-30. Rasterizing the pdf is clearly not the way to go if you simply have a pdf with terrible OCR that you want to redo.

Recently, I have begun fixing terrible OCR in documents by simply running ```gs -o out.pdf -sDEVICE=pdfwrite -dFILTERTEXT in.pdf``` over the document first, and then running ocrmypdf without either ```-redo-ocr``` or ```-f```.

Obviously, this is not something that necessarily needs to be included in ocrmypdf itself. On the other hand, just a couple of google searches have shown me that I am not the only one who is unhappy with ```-f``` and ```--redo-ocr``` though, and there are no helpful suggestions for people on stackexchange. So maybe it would be nice to advertise this option by including it in ocrmypdf as a flag.

Here is an example of a file that cannot be salvaged with ```-f``` or ```--redo-ocr```.  https://www.jstor.org/stable/4134353 . Sorry if it is behind a paywall for you, use sci-hub to get around that. The only problem with that file is that the original ocr blends the two columns into one. ```-f``` blows up the file size from 1.2 MB to ~30. ```--redo-ocr``` scrambles the text by adding additional text on top of the existing text. As a side note, for something like jstor the new option of removing the text layer and running ocr would play nicely with the ```--pages``` argument in case it does not work for all pages.

**Comment by jbarlow83** [comment]
OCRmyPDF is geared towards batch processing jobs where review of all PDF processed is unlikely if not impossible. Thus it is strongly biased towards caution and preserving data. Better a big file with OCR and all your original data than a damaged one.

`--force-ocr` increases file sizes in case where the PDF contains color content that forces promotion to full color and rendering of vector graphics including fonts to bitmap. We don't have a color segmentation algorithm, which would improve the situation. There are other parameters that tune output size, notably optimization.

Some older versions of OCRmyPDF promoted to full color in cases where it was not necessarily, but no version of ocrmypdf was indicated.

`gs -dFILTERTEXT` removes both OCR (invisible text) and printable text without discrimination - so it can permanently remove visible content from files without any warning, making it unsuitable for general use even though it does work on files with OCR. There is no standard way to distinguish OCR text from printable text, so `--redo-ocr` does not work on all ways of encoding OCR, unfortunately. This is a documented limitation.

Tesseract seems to correctly recognize multi column text but may not order the output correctly in its PDF output renderer. However, some of these behaviors are necessarily heuristic and vary among PDF viewers. This needs investigation. (PDF/A conversion is not the issue.)

**Comment by Jmuccigr** [comment]
See #736 with lots of comments, including mine about how I handle files that I need to redo the OCR on:

> I have this same thing happen all the time when I redo the OCR on a PDF by removing the text layer and using ocrmypdf --force-ocr on the file, then overlaying just the text layer back onto the original file (I use --output-type pdf not pdfa), or even when I use tesseract to OCR some image files and put them into a PDF.

When I've got a PDF that is just OCRed images, I use PyPDF2 to remove the text and save that to a new file. Then ocrmypdf with force-ocr OCRs that new file, saving to a second file on which I have gs remove all the images leaving just text, and finally(!) qpdf to overlay the text layer back onto the original pdf. I save some of the original file's metadata and also have an option to skip the first page (because I do this on jstor files a lot and I like to keep that info page). Script is [here](https://github.com/Jmuccigr/scripts/blob/master/redo_ocr_PDF.sh).

**Comment by julianbarg** [comment]
@jbarlow83 That is correct, apart from the file size increase, ocrmypdf works great. Unfortunately, rasterizing can go really bad in terms of file size. I have not done enough experimentation, but I also vaguely remember seeing some bad interaction with pdfs where the file size was set incorrectly (e.g., books blown up to 3-4 times their original size).

@Jmuccigr Thank you fo the great advice, that is a good approach.

**Comment by Mark-Joy** [comment]
> `gs -dFILTERTEXT` removes both OCR (invisible text) and printable text without discrimination - so it can permanently remove visible content from files without any warning, making it unsuitable for general use even though it does work on files with OCR. There is no standard way to distinguish OCR text from printable text, so `--redo-ocr` does not work on all ways of encoding OCR, unfortunately. This is a documented limitation.
Is there a way to distinguish visible text and invisible text based on its opaque/transparency property?
In case there is, we can safely remove only invisible text and then `--redo-ocr`. The result pdf would be the accuracy of `--force-ocr` without rasterizing text.

**Comment by jbarlow83** [comment]
Not easily. Some OCR engines encode OCR text as visible, sometimes with some sort of markup to indicate OCR, and paint over it. More information on the variety of approaches using by various software would be helpful - I don't know if that's been gathered by anyone.

**Comment by Mark-Joy** [comment]
I am only an user with very little knowledge about pdf. So I'm sorry for being intrusive to propose the below logic if it is not already implemented:

1. We have `OCR_remove` as a function to remove OCR text. Using feedback/user report here, this function can be improved over time by adding more methods to detect OCR text.
2. Original Pdf as `PDF0`
	`PDF0`: perform `OCR_remove` to remove only OCR text -> `PDF1`
	`PDF1`: remove ALL text (including visible/printable text) -> `PDF2`
	Perform `OCR` on `PDF2` -> we get `Result_OCRText`
	Combine (`PDF1` + `Result_OCRText`) to `Final PDF`.

**Comment by jbarlow83** [comment]
ocrmypdf more or less implements what you describe in --redo-ocr, although the PDF1->PDF2 step is done using images, and we remove text by masking the image we send out for OCR. The difficulty is in removing OCR text correctly.

See `ocrmypdf.pdfinfo.layout, ocrmypdf._graft.strip_invisible_text, ocrmypdf._pipeline.create_ocr_image`
